src/hotspot/cpu/s390/templateTable_s390.cpp

```diff
@@ -2342,57 +2355,111 @@ void TemplateTable::_return(TosState state) {
+  __ load_const_optimized(Z_ARG2, (int)code);
   __ call_VM(noreg, entry, Z_ARG2);
-
   // Update registers with resolved info.
-  __ get_cache_and_index_at_bcp(cache, cpe_offset, 1, index_size);
+  __ get_cache_and_index_at_bcp(cache, index, 1, index_size);
+
   __ bind(resolved);
 
   // Class initialization barrier for static methods
   if (VM_Version::supports_fast_class_init_checks() && bytecode() == Bytecodes::_invokestatic) {
     const Register method = Z_R1_scratch;
     const Register klass  = Z_R1_scratch;
 
-    __ load_resolved_method_at_index(byte_no, cache, cpe_offset, method);
+    __ load_resolved_method_at_index(byte_no, cache, index, method);
     __ load_method_holder(klass, method);
     __ clinit_barrier(klass, Z_thread, nullptr /*L_fast_path*/, &clinit_barrier_slow);
   }
 
   BLOCK_COMMENT("} resolve_cache_and_index");
 }
 
+void TemplateTable::resolve_cache_and_index_for_field(int byte_no,
+                                                      Register cache,
+                                                      Register index) {
+
+  assert_different_registers(cache, index);
+  assert(byte_no == f1_byte || byte_no == f2_byte, "byte_no out of range");
+
+  NearLabel resolved;
+
+  Bytecodes::Code code = bytecode();
+  switch (code) {
+    case Bytecodes::_nofast_getfield: code = Bytecodes::_getfield; break;
+    case Bytecodes::_nofast_putfield: code = Bytecodes::_putfield; break;
+    default: break;
+  }
+
+  __ load_field_entry(cache, index);
+  if (byte_no == f1_byte) {
+    __ z_cli(Address(cache, in_bytes(ResolvedFieldEntry::get_code_offset())), code);
+  } else {
+    __ z_cli(Address(cache, in_bytes(ResolvedFieldEntry::put_code_offset())), code);
+  }
+  __ z_bre(resolved);
+
+  // resolve first time through
+  address entry = CAST_FROM_FN_PTR(address, InterpreterRuntime::resolve_from_cache);
+  __ load_const_optimized(Z_ARG2, (int)code);
+  __ call_VM(noreg, entry, Z_ARG2);
+
+  // Update registers with resolved info.
+  __ load_field_entry(cache, index);
+
+  __ bind(resolved);
+}
+
+// The cache register (the only input reg) must be set before call.
+void TemplateTable::load_resolved_field_entry(Register obj,
+                                              Register cache,
+                                              Register tos_state,
+                                              Register offset,
+                                              Register flags,
+                                              bool is_static = false) {
+  assert_different_registers(cache, tos_state, flags, offset);
+
+  // Field offset
+  __ load_sized_value(offset, Address(cache, in_bytes(ResolvedFieldEntry::field_offset_offset())), sizeof(int), true /*is_signed*/);
+
+  // Flags
+  __ load_sized_value(flags, Address(cache, in_bytes(ResolvedFieldEntry::flags_offset())), sizeof(u1), false);
+
+  // TOS state
+  __ load_sized_value(tos_state, Address(cache, in_bytes(ResolvedFieldEntry::type_offset())), sizeof(u1), false);
```

Comment:
Note that tos_state is not needed by all usages. Making this optional saves an instruction and a register (see PPC64 implementation).